test/unit_test/runtime/linear_algebra_test.cpp

```diff
@@ -0,0 +1,289 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <linear_algebra.hpp>
+// linear_algebra.hpp has to be included first otherwise the header will fail to compile!
+#include <catch2/catch_test_macros.hpp>
+#include <mp_units/customization_points.h>
+#include <mp_units/format.h>
+#include <mp_units/quantity_io.h>
+#include <mp_units/systems/isq/mechanics.h>
+#include <mp_units/systems/isq/space_and_time.h>
+#include <mp_units/systems/si/si.h>
+#include <cmath>
+
+namespace STD_LA {
+
+template<class ET, class OT>
+std::ostream& operator<<(std::ostream& os, const vector<ET, OT>& v)
+{
+  os << "|";
+  for (auto i = 0U; i < v.size(); ++i) {
+    os << STD_FMT::format(" {:>9}", v(i));
+  }
+  os << " |";
+  return os;
+}
+
+template<class ET, class OT>
+std::ostream& operator<<(std::ostream& os, const matrix<ET, OT>& v)
+{
+  for (auto i = 0U; i < v.rows(); ++i) {
+    os << "|";
+    for (auto j = 0U; j < v.columns(); ++j) {
+      os << STD_FMT::format(" {:>9}", v(i, j));
+    }
+    os << (i != v.rows() - 1U ? " |\n" : " |");
+  }
+  return os;
+}
+
+}  // namespace STD_LA
+
+using namespace mp_units;
+using namespace mp_units::si::unit_symbols;
+
+template<typename Rep = double>
+using vector = std::math::fs_vector<Rep, 3>;
+
+template<typename Rep>
+inline constexpr bool mp_units::is_vector<vector<Rep>> = true;
+
+namespace {
+
+template<typename T>
+[[nodiscard]] auto get_magnitude(const vector<T>& v)
+{
+  return std::hypot(v[0], v[1], v[2]);
+}
+
+template<typename T, typename U>
+[[nodiscard]] vector<decltype(T{} * U{})> cross_product(const vector<T>& a, const vector<U>& b)
+{
+  return {a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]};
+}
+
+template<Quantity Q1, Quantity Q2>
+  requires is_vector<typename Q1::rep> && is_vector<typename Q2::rep> &&
+           requires(typename Q1::rep v1, typename Q2::rep v2) { cross_product(v1, v2); }
+[[nodiscard]] quantity_of<Q1::reference * Q2::reference> auto cross_product(const Q1& q1, const Q2& q2)
+{
+  return (Q1::reference * Q2::reference)(cross_product(q1.number(), q2.number()));
+}
+
+}  // namespace
+
+TEST_CASE("vector quantity", "[la]")
+{
+  SECTION("cast of unit")
+  {
+    SECTION("non-truncating")
+    {
+      const quantity<isq::position_vector[km], vector<int>> v{vector<int>{3, 2, 1}};
+      CHECK(v[m].number() == vector<int>{3000, 2000, 1000});
+    }
+
+    SECTION("truncating")
+    {
+      const quantity<isq::position_vector[m], vector<int>> v{vector<int>{1001, 1002, 1003}};
+      CHECK(quantity_cast<km>(v).number() == vector<int>{1, 1, 1});
+    }
+  }
+
+  SECTION("to scalar magnitude")
+  {
+    const quantity<isq::velocity[km / h], vector<int>> v{vector<int>{2, 3, 6}};
+    const auto speed = get_magnitude(v.number()) * isq::speed[v.unit];  // TODO can we do better here?
```

Comment:
One of my ideas with
```C++
inline constexpr auto velocity     = defn<"𝙫", "𝙫 = d𝙧/d𝘵", position_vector("𝙧"), duration("𝘵")>();
inline constexpr auto speed        = defn<"𝘷", "𝘷 = |𝙫|", velocity("𝙫")>();
```
was to define `speed` as a magnitude of `velocity`. By some customization point, `isq::speed(v)` could be made to work.

Reply:
I agree. I think that `get_magnitude(v)` should return a `quantity` with a `quantity_spec` being something like `derived_quantity_spec<magnitude<isq::velocity>>` and `isq::speed` should be exactly defined as such. For example:

```cpp
inline constexpr struct speed : quantity_spec<get_magnitude(velocity)> {} speed;
```

This will allow the implicit conversion from an unnamed `derived_quantity_spec<...>` to a named `speed`.

Reply:
And well, this name somehow collides with the `magnitude` in the unit definition. I think that the unit's magnitude engine is overcomplicated and could be highly simplified, maybe even up to the point where we will not need to use a `magnitude` name anymore.